### DISCUSSION TITLE
How do I get a Twitch OAuth token to authenticate my requests?

### DISCUSSION THREAD

**Discussion by serg06** [streamlink/streamlink#4400] [Q&A] [answered]
The docs list a CLI option [--twitch-api-header](https://streamlink.github.io/cli.html?highlight=twitch%20api%20header#cmdoption-twitch-api-header).

In commit dfea62af0fd8cc2eeb9781e46583c613214a7c2c, it notes "- can be used for `Authorization` or other headers".

I tried getting an OAuth token from [here](https://twitchtokengenerator.com/), and even though the token works for the regular API,

![image](https://user-images.githubusercontent.com/1350889/158681577-75e9459f-6bcf-4f02-999d-1e027accd231.png)

it doesn't work for Streamlink.

![image](https://user-images.githubusercontent.com/1350889/158681602-24ae794f-fc58-49c8-bd5b-bfe33e17c6d7.png)

If I had to guess, it's because Streamlink uses the GQL API:

https://github.com/streamlink/streamlink/blob/a8b2ddcd8629f907363992479ba3065f96def261/src/streamlink/plugins/twitch.py#L230

Can anyone explain how to get an OAuth token for the GQL API? @back-to perhaps you know, since you created that feature?

**Comment by bastimeyer** [comment, marked as answer]
> it's because Streamlink uses the GQL API

This is indeed meant for the GraphQL API which is used by Twitch's website and which is needed to retrieve the streaming access tokens. The Helix REST API is something entirely else.

OAuth tokens for the GQL API can only be retrieved via your web browser's dev tools as 3rd party websites/tools can't authenticate users on it. Log in with your account on Twitch.tv, open the dev tools, go to the networks tab and check for any request being made to `gql.twitch.tv`. Any of those requests will have an `Authorization` request header if you're logged in. This is what Streamlink's `--twitch-api-token` CLI parameter is for (in addition to other HTTP headers related to ads). The reason for this is that if HTTP headers get set via Streamlink's `--http-header "key=value"` CLI parameter, they get set on the entire HTTP session, which includes HLS streams, and we don't want that. `--twitch-api-token` is a plugin parameter and it only sets the headers on gql.twitch.tv API requests.

https://github.com/streamlink/streamlink/blob/3.2.0/src/streamlink/plugins/twitch.py#L219-L225

1. Log in on Twitch.tv in your web browser
2. Press F12, go to the networks tab
3. Click on any of the `gql.twitch.tv` HTTP requests
4. Copy the value of the `Authorization` request header, which is `OAuth YOUROAUTHTOKEN`
5. Set the `--twitch-api-header "Authorization=OAuth YOUROAUTHTOKEN"` CLI parameter+value

**Reply by serg06** [reply]
> OAuth tokens for the GQL API can only be retrieved via your web browser's dev tools as 3rd party websites/tools can't authenticate users on it.

Damn! I guess if I want to automate it, I'll have to launch a browser instance. Oh well, thank you!
